spd/utils/component_utils.py

```diff
@@ -22,24 +22,87 @@ def _sample_stochastic_mask(
     return causal_importances + (1 - causal_importances) * rand_tensor
 
 
+RoutingType = Literal["uniform_k-stochastic", "all"]
+"""How to choose which (batch,) or (batch, seq_len) positions to route to components or target.
+
+uniform_k-stochastic:
+    for each position, sample k from [1, n_modules], then route to components for k out of `n_modules` modules
+all:
+    use components for all positions
+"""
+
+
+def rand_perm(shape: tuple[int, ...], dim: int, device: torch.device | str) -> Int[Tensor, "... k"]:
+    noise = torch.rand(shape, device=device)
+    # turn values into ranks via double argsort trick. (for example: [0.8, 0.2, 0.3] -> [2, 0, 1])
+    return noise.argsort(dim=dim).argsort(dim=dim)
+
+
+def sample_uniform_k_subset_routing_masks(
```

Comment:
I think this guy could do with a couple of unittests. Just some that show that we get samples in the range that we'd expect for (batch,) and (batch, seq).